Release 1.2 (#37)
ingonyama-zk/developer-docs: docs/icicle/primitives/poseidon.md

# Poseidon

> Poseidon hash is a popular hash in the ZK ecosystem primarily because its optimized to work over large prime fields, a common setting for ZK proofs, thereby minimizing the number of multiplicative operations required.

Reviewer: Actually, I thought that the main feature is it's ability to easily fit inside ZK circuits. So computing Poseidon `inside` a ZK circuit is much much faster than Keccak or Sha-256. Maybe I am wrong

Author: its also that.

Author: Both points are very related, ill add a note about ur point as well

#### Partial Rounds

**Selective Application of S-Box:** Partial rounds apply the S-box transformation to only one element of the internal state per round, rather than to all elements. This selective application significantly reduces the computational complexity of the hash function without compromising its security. The choice of which element to apply the S-box to can follow a specific pattern or be fixed, depending on the design of the hash function.

**Linear Transformation and Round Constants:** A linear transformation is performed, and round constants are added. However, the linear transformation in partial rounds can be designed to be less computationally intensive (this is done by using a sparse matrix) than in full rounds, further optimizing the function's efficiency.

The user of a Poseidon hash often can choose how many partial or full rounds he wishes to apply, more full rounds will increase security and degrade performance. The choice and balance depend highly on the use case.

![Alt text](image-2.png)

## Using Poseidon Hash

> ICICLE Poseidon hash is implemented for GPU and parallelization is performed for each element of the state rather than for each state.

Reviewer: It's important to note that our implementation is focused on calculating multiple hash-sums over multiple pre-images in parallel. Rather than going block by block over the input vector.

So for Poseidon of arity 2 and input of size 1024 * 2, you will get 1024 elements of output. Which is the result of calculating 1024 Poseidon hashes over blocks of size 2.

Author: Good point, I copied ur explanation into the docs sine i think its simple and clear.

### Supported API

[`Rust`](https://github.com/ingonyama-zk/icicle/tree/main/wrappers/rust/icicle-core/src/poseidon), [`C++`](https://github.com/ingonyama-zk/icicle/tree/main/icicle/appUtils/poseidon)

### Supported curves

Poseidon supports the following curves:

`bls12-377`, `bls12-381`, `bn-254`, `bw6-761`

### Constants

Poseidon is extremely customizable using different constants will producing different hashes, different security levels and different performance results.

We support pre calculated and optimized constants for each of the [supported curves](#supported-curves).The constants can be found [here](https://github.com/ingonyama-zk/icicle/tree/main/icicle/appUtils/poseidon/constants), the are labeled clearly per curve `<curve_name>_poseidon.h`.

If you wish to generate your own constants you can use our python script which can be found [here](https://github.com/ingonyama-zk/icicle/blob/b6dded89cdef18348a5d4e2748b71ce4211c63ad/icicle/appUtils/poseidon/constants/generate_parameters.py#L1).

Prerequisites:
- Install python 3
- `pip install poseidon-hash`

> - `pip install galois`

Reviewer: `pip install galois==0.3.7`

This is crucial

Author: done

            are_outputs_on_device: false,
            input_is_a_state: false,
            aligned: false,
            loop_state: false,
            is_async: false,
        }
    }
}
```

In the example above `load_optimized_poseidon_constants::<F>(arity, &ctx).unwrap();` is used which will load the correct constants based on arity and curve. Its possible to [generate](#constants) your own constants and load them.

```rust
let ctx = get_default_device_context();
    let cargo_manifest_dir = env!("CARGO_MANIFEST_DIR");
    let constants_file = PathBuf::from(cargo_manifest_dir)
        .join("tests")
        .join(format!("{}_constants.bin", field_prefix));
    let mut constants_buf = vec![];
    File::open(constants_file)
        .unwrap()
        .read_to_end(&mut constants_buf)
        .unwrap();

    let mut custom_constants = vec![];
    for chunk in constants_buf.chunks(field_bytes) {
        custom_constants.push(F::from_bytes_le(chunk));
    }

    let custom_constants = create_optimized_poseidon_constants::<F>(
        arity as u32,
        &ctx,
        full_rounds_half,
        partial_rounds,
        &mut custom_constants,
    )
    .unwrap();
```

> For more examples using different configurations refer here.

Reviewer: refer where?

Author: the examples pr isnt merged yet i have a todo to update this.

### Benchmarks 

TODO

> ## The Tree Builder

Reviewer: Shouldn't it be on a different page?

Author: For now I think this is good, as we expand the tree builder more we will end up giving it a dedicated page.

# Poseidon

> Poseidon hash is a popular hash in the ZK ecosystem primarily because its optimized to work over large prime fields, a common setting for ZK proofs, thereby minimizing the number of multiplicative operations required.

Reviewer 2: ```suggestion
Poseidon hash is a popular hashing algorithm in the ZK ecosystem primarily because it is optimized to work over large prime fields, a common setting for ZK proofs, thereby minimizing the number of multiplicative operations required.
```

> Poseidon hash function has also been specifically designed to be efficient when implemented within ZK circuits, Poseidon hash uses far less constraints compared to other hash functions like Keccak or SHA-256 in the context of ZK circuits.

Reviewer 2: I think it makes sense to refer to Poseidon as strictly "Poseidon" and not "Poseidon hash" or "Poseidon hash function" after the first sentence describing it as a hash function

Poseidon hash has been used in many popular ZK protocols such as Filecoin and [Plonk](https://drive.google.com/file/d/1bZZvKMQHaZGA4L9eZhupQLyGINkkFG_b/view?usp=drive_open).

> Our implementation of [Poseidon hash](https://eprint.iacr.org/2019/458.pdf) is implemented  in accordance with the optimized [Filecoin version](https://spec.filecoin.io/algorithms/crypto/poseidon/).

Reviewer 2: I would move the link to the Poseidon paper to an earlier (if not the first) mention of Poseidon hash. Its especially odd here where we are saying "Our implementation" a the beginning of the sentence and not linking to our implementation

Author: agree, thanks

> Our implementation of [Poseidon hash](https://eprint.iacr.org/2019/458.pdf) is implemented  in accordance with the optimized [Filecoin version](https://spec.filecoin.io/algorithms/crypto/poseidon/).

Reviewer 2: ```suggestion
Our implementation of [Poseidon hash](https://eprint.iacr.org/2019/458.pdf) is implemented in accordance with the optimized [Filecoin version](https://spec.filecoin.io/algorithms/crypto/poseidon/).
```

> Let understand how Poseidon hash works.

Reviewer 2: ```suggestion
Let's understand how Poseidon hash works.
```

### Initialization

Poseidon starts with the initialization of its internal state, which is composed of the input elements and some pregenerated constants. An initial round constant is added to each element of the internal state. This is part of the initialization to ensure the state is properly mixed from the outset.

> This is done to prevent collions and to prevent certain cryptographic attacks by insuring that the internal state is sufficiently mixed and unpredictable,

Reviewer 2: What is "this" referring to on line 17? Mixing?

Author: yes fixed the writing

> This is done to prevent collions and to prevent certain cryptographic attacks by insuring that the internal state is sufficiently mixed and unpredictable,

Reviewer 2: ```suggestion
This is done to prevent collisions and to prevent certain cryptographic attacks by ensuring that the internal state is sufficiently mixed and unpredictable.
```

![Alt text](image.png)

### Applying full and partial rounds

> To generate a secure hash output the algorithm goes through a series of "full round" and "partial rounds" as well as transformations between these sets of rounds.

Reviewer 2: ```suggestion
To generate a secure hash output, the algorithm goes through a series of "full rounds" and "partial rounds" as well as transformations between these sets of rounds.
```

> First full rounds => apply SBox and Round constants => partial rounds => Last full rounds => Apply SBox

Reviewer 2: Maybe it makes sense to have this as some type of visual instead of text?

Author: I have the images. Do u have any suggestion ?

#### Full rounds

![Alt text](image-1.png)

> **Uniform Application of S-Box:** In full rounds, the S-box (a non-linear transformation) is applied uniformly to every element of the hash function's internal state. This ensures a high degree of mixing and diffusion, contributing to the hash function's security. The S-box involves raising each element of the state to a certain power denoted by `α` is a member of the finite field defined by the prime `p`, `α` can be selected by the user.

Reviewer 2: "The S-box involves raising each element of the state to a certain power denoted by `α` is a member of the finite field defined by the prime `p`, `α` can be selected by the user."

I think there is something missing here "...denoted by `α` is a member of the...."

**Linear Transformation:** After applying the S-box, a linear transformation is performed on the state. This involves multiplying the state by a MDS (Maximum Distance Separable) Matrix. which further diffuses the transformations applied by the S-box across the entire state.

**Addition of Round Constants:** Each element of the state is then modified by adding a unique round constant. These constants are different for each round and are precomputed as part of the hash function's initialization. The addition of round constants ensures that even minor changes to the input produce significant differences in the output.

#### Partial Rounds

**Selective Application of S-Box:** Partial rounds apply the S-box transformation to only one element of the internal state per round, rather than to all elements. This selective application significantly reduces the computational complexity of the hash function without compromising its security. The choice of which element to apply the S-box to can follow a specific pattern or be fixed, depending on the design of the hash function.

> **Linear Transformation and Round Constants:** A linear transformation is performed, and round constants are added. However, the linear transformation in partial rounds can be designed to be less computationally intensive (this is done by using a sparse matrix) than in full rounds, further optimizing the function's efficiency.

Reviewer 2: ```suggestion
**Linear Transformation and Round Constants:** A linear transformation is performed and round constants are added. The linear transformation in partial rounds can be designed to be less computationally intensive (this is done by using a sparse matrix) than in full rounds, further optimizing the function's efficiency.
```

> The user of a Poseidon hash often can choose how many partial or full rounds he wishes to apply, more full rounds will increase security and degrade performance. The choice and balance depend highly on the use case.

Reviewer 2: ```suggestion
The user of a Poseidon hash can often choose how many partial or full rounds he wishes to apply; more full rounds will increase security but degrade performance. The choice and balance is highly dependent on the use case.
```

![Alt text](image-2.png)

## Using Poseidon Hash

ICICLE Poseidon hash is implemented for GPU and parallelization is performed for each element of the state rather than for each state.
What that means is we calculate multiple hash-sums over multiple pre-images in parallel, rather than going block by block over the input vector.

So for Poseidon of arity 2 and input of size 1024 * 2, we would expect 1024 elements of output. Which means each block would be of size 2 and that would result in 1024 Poseidon hashes being performed.

### Supported API

[`Rust`](https://github.com/ingonyama-zk/icicle/tree/main/wrappers/rust/icicle-core/src/poseidon), [`C++`](https://github.com/ingonyama-zk/icicle/tree/main/icicle/appUtils/poseidon)

### Supported curves

Poseidon supports the following curves:

`bls12-377`, `bls12-381`, `bn-254`, `bw6-761`

### Constants

> Poseidon is extremely customizable using different constants will producing different hashes, different security levels and different performance results.

Reviewer 2: ```suggestion
Poseidon is extremely customizable and using different constants will produce different hashes, security levels and performance results.
```

> We support pre calculated and optimized constants for each of the [supported curves](#supported-curves).The constants can be found [here](https://github.com/ingonyama-zk/icicle/tree/main/icicle/appUtils/poseidon/constants), the are labeled clearly per curve `<curve_name>_poseidon.h`.

Reviewer 2: ```suggestion
We support pre-calculated and optimized constants for each of the [supported curves](#supported-curves).The constants can be found [here](https://github.com/ingonyama-zk/icicle/tree/main/icicle/appUtils/poseidon/constants) and are labeled clearly per curve `<curve_name>_poseidon.h`.
```

            aligned: false,
            loop_state: false,
            is_async: false,
        }
    }
}
```

In the example above `load_optimized_poseidon_constants::<F>(arity, &ctx).unwrap();` is used which will load the correct constants based on arity and curve. Its possible to [generate](#constants) your own constants and load them.

```rust
let ctx = get_default_device_context();
    let cargo_manifest_dir = env!("CARGO_MANIFEST_DIR");
    let constants_file = PathBuf::from(cargo_manifest_dir)
        .join("tests")
        .join(format!("{}_constants.bin", field_prefix));
    let mut constants_buf = vec![];
    File::open(constants_file)
        .unwrap()
        .read_to_end(&mut constants_buf)
        .unwrap();

    let mut custom_constants = vec![];
    for chunk in constants_buf.chunks(field_bytes) {
        custom_constants.push(F::from_bytes_le(chunk));
    }

    let custom_constants = create_optimized_poseidon_constants::<F>(
        arity as u32,
        &ctx,
        full_rounds_half,
        partial_rounds,
        &mut custom_constants,
    )
    .unwrap();
```
For more examples using different configurations refer here.

### Benchmarks 

> TODO

Reviewer 2: Add or remove
